controller/object_store_controller.go

```diff
@@ -0,0 +1,1159 @@
+	if err == nil {
+		return pv, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		pv, err = osc.createPV(store, volume)
+		if err != nil {
+			return nil, store, errors.Wrap(err, "failed to create persistent volume")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return pv, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkPV(pv *corev1.PersistentVolume) error {
+	if pv.Status.Phase != corev1.VolumeBound {
+		return errors.New(fmt.Sprintf("PV %v not bound", pv.Name))
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) getOrCreateDeployment(store *longhorn.ObjectStore) (*appsv1.Deployment, *longhorn.ObjectStore, error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err == nil {
+		return dpl, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		dpl, err = osc.createDeployment(store)
+		if err != nil {
+			return nil, store, errors.Wrap(err, "failed to create deployment")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return dpl, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkDeployment(deployment *appsv1.Deployment, store *longhorn.ObjectStore) error {
+	if *deployment.Spec.Replicas != 1 {
+		deployment.Spec.Replicas = int32Ptr(1)
+		osc.ds.UpdateDeployment(deployment)
+		return errors.New("deployment just scaled")
+	} else if deployment.Status.Replicas == 0 || deployment.Status.UnavailableReplicas > 0 {
+		return errors.New("deployment not ready")
+	}
+
+	if store.Spec.Image != "" && deployment.Spec.Template.Spec.Containers[0].Image != store.Spec.Image {
+		deployment.Spec.Template.Spec.Containers[0].Image = store.Spec.Image
+		osc.ds.UpdateDeployment(deployment)
+		store.Status.State = longhorn.ObjectStoreStateStarting
+	}
+
+	if store.Spec.UiImage != "" && deployment.Spec.Template.Spec.Containers[1].Image != store.Spec.UiImage {
+		deployment.Spec.Template.Spec.Containers[1].Image = store.Spec.UiImage
+		osc.ds.UpdateDeployment(deployment)
+		store.Status.State = longhorn.ObjectStoreStateStarting
+	}
+
+	return nil
+}
+
+func (osc *ObjectStoreController) getOrCreateService(store *longhorn.ObjectStore) (*corev1.Service, *longhorn.ObjectStore, error) {
+	svc, err := osc.ds.GetService(osc.namespace, store.Name)
+	if err == nil {
+		return svc, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		svc, err = osc.createService(store)
+		if err != nil {
+			return nil, store, errors.Wrap(err, "failed to create service")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return svc, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) getOrCreateS3Endpoints(store *longhorn.ObjectStore) ([]*networkingv1.Ingress, *longhorn.ObjectStore, error) {
```

Comment:
NIT: no need to output `*longhorn.ObjectStore`

Reply:
This function updates the store's `.status.endpoints` array when it creates the ingress resources for each endpoint. This change needs to be carried forward until we call `ds.UpdateObjectStoreStatus` at the exit of the `reconcile` function just like any other change to the `.status` property.